File: src/main/java/com/github/tianchenghang/agent/ReActAgent.java
package com.github.tianchenghang.agent;

import cn.hutool.core.util.StrUtil;
import java.util.HashMap;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.extern.slf4j.Slf4j;

// ReAct (Reasoning and Acting)
@EqualsAndHashCode(callSuper = true)
@Data
@Slf4j
public abstract class ReActAgent extends BaseAgent {
  public abstract boolean think();

  public abstract String act();

  protected void emitThinking(String content) {
    if (StrUtil.isNotBlank(content)) {
      sendSseMessage(EventNames.THINKING, content);
    }
  }

  protected void emitToolCall(String toolName, String args) {
    var data = new HashMap<String, Object>();
    data.put("tool", toolName);
    data.put("args", args);
    sendSseMessage(EventNames.TOOL_CALL, data);
  }

  protected void emitToolResult(String toolName, String result) {
    var data = new HashMap<String, Object>();
    data.put("tool", toolName);
    data.put("result", result);
    sendSseMessage(EventNames.TOOL_RESULT, data);
  }

  @Override
  public String step() {
    try {
      var shouldAct = think();
      if (!shouldAct) {
        return "Thinking complete, no action required";
      }
      return act();
    } catch (Exception e) {
      e.printStackTrace();
      return "Step execute failed: " + e.getMessage();
    }
  }
}

File: server/src/agent/re-act-agent.ts
import { Subject } from 'rxjs';
import { BaseAgent } from './base-agent';
import { EventName } from './model/event-name.enum';

export abstract class ReActAgent extends BaseAgent {
  protected emitter: Subject<string> | null = null;

  abstract think(): Promise<boolean>;

  abstract act(): Promise<string>;

  // Send think content to SSE (Server-Sent Events)
  protected emitThinking(content: string) {
    if (this.emitter && content.trim()) {
      this.emitter.next(JSON.stringify({ type: EventName.THINKING, content }));
    }
  }

  // Send tool call to SSE (Server-Sent Events)
  protected emitToolCall(toolName: string, args: string) {
    if (this.emitter) {
      this.emitter.next(
        JSON.stringify({ type: EventName.TOOL_CALL, tool: toolName, args }),
      );
    }
  }

  // Send tool result to SSE (Server-Sent Events)
  protected emitToolResult(toolName: string, result: string) {
    if (this.emitter) {
      this.emitter.next(
        JSON.stringify({ type: EventName.TOOL_RESULT, tool: toolName, result }),
      );
    }
  }

  override async step(): Promise<string> {
    try {
      const shouldAct = await this.think();
      if (!shouldAct) {
        return 'Thinking complete, no action required';
      }
      return await this.act();
    } catch (err) {
      this.logger.log('Step execute error:', err);
      return 'Step execute error';
    }
  }

  protected override async stepWithEmitter(
    emitter: Subject<string>,
  ): Promise<string> {
    this.emitter = emitter;
    try {
      return await this.step();
    } finally {
      this.emitter = null;
    }
  }
}
